File: Searching/BinarySearch.py
# def BinarySearch(arr,target):
# 	arr.sort()
# 	l=0
# 	r=len(arr)-1
# 	while(l<=r):
# 		mid = (l+r)//2
# 		if(arr[mid] == target):
# 			print("Found Element",arr[mid],"At index",mid)
# 			return
# 		elif(arr[mid] > target):
# 			r = mid-1
# 		else:
# 			l = mid+1
# 	print("Not Found")






def BinarySearch(arr,l,r,target):
	if(l>r):
		print("Not Found")
		return
	else:
		mid = (l+r)//2
		if(arr[mid] == target):
			print("Found Element",arr[mid],"At index",mid)
			return 
		elif(arr[mid] > target):
			BinarySearch(arr,l,mid-1,target)
		else:
			BinarySearch(arr,mid+1,r,target)




if __name__ == '__main__':
	target = int(input("Enter a element to be searched =>"))
	arr = [24,6,58,65,95,9]
	arr.sort()
	print(arr)
	l=0
	r=len(arr)-1
	BinarySearch(arr,l,r,target)

File: 6-Searching/BinarySearch.cpp
#include<bits/stdc++.h>
using namespace std;
typedef long long ll;


ll binarySearch(vector<ll> &v,int l,int j,int target){
	if(l>j){
		return -1;
	}

	ll mid = (l+j)/2;
	if(target == v[mid]){
		return mid;
	}else if(v[mid] < target){
		binarySearch(v,mid+1,j,target);
	}else{
		binarySearch(v,l,mid-1,target);
	}

}


int main()
{
	ll t,n,i,j,q;
	cin>>n;
	vector<ll> v;
	for(int i=0;i<n;i++){
		cin>>q;
		v.push_back(q);
	}
	sort(v.begin(), v.end());
	ll target = 0;
	cin>>target;
	ll res = binarySearch(v,0,v.size()-1,target);
	if(res == -1){

		cout<<"Not Found "<<res;
	}else{
		cout<<"Found At "<<res;
	}

	return 0;
}